lib/cb/utils/validator.rb

```diff
@@ -23,8 +23,19 @@ def validate(response)
 
       def raise_response_code_errors(response)
         code = response.code rescue nil
-        fail Cb::ServiceUnavailableError if code == 503 || simulate_auth_outage?
-        fail Cb::UnauthorizedError if code == 401
+        fail_with_error_details(response, Cb::ServiceUnavailableError) if code == 503 || simulate_auth_outage?
+        fail_with_error_details(response, Cb::UnauthorizedError) if code == 401 || code == 403
+        fail_with_error_details(response, Cb::DocumentNotFoundError) if code == 404
+        fail_with_error_details(response, Cb::BadRequestError) if code && code >= 400 && code < 500
+        fail_with_error_details(response, Cb::ServerError) if code && code >= 500
```

Comment:
Was an entertaining thought experiment to figure out how I would do this method differently, here's what I came up with:

```ruby
def raise_response_code_errors(response)
  code = response.respond_to?(:code) ? response.code : nil
  return unless code
  error_klass = nil
  if code >= 500
    error_klass = ((code == 503 || simulate_auth_outage?) ? Cb::ServiceUnavailableError : Cb::ServerError)
  elsif code >= 400
    error_klass = Cb::BadRequestError
    error_klass = (code == 404 ? Cb::DocumentNotFoundError : Cb::UnauthorizedError) if [401, 403, 404].include?(code)
  end
  fail_with_error_details(response, error_klass) if error_klass
end
```

The main things that bothered me with the version you have is the code duplication with each line doing `fail_with_error_details` and the `rescue nil` on the `response.code` line. However, the reality is that your attack was fewer lines of code so it ends up being 6 of one, half dozen of another.